docs/USAGE-GUIDE.md

```diff
@@ -1373,6 +1382,72 @@ int s2n_config_add_ticket_crypto_key(struct s2n_config *config, const uint8_t *n
+executed each time operation on private key needs to be performed with
+**op** object representing this operation. From this function the user
+can spawn the thread with **op** object to perform through
+**s2n_async_pkey_op_perform** call and immediately return from the
+function without waiting for thread to complete. The **s2n_negotiate**
+will fail with **S2N_ERR_T_BLOCKED** error type and **s2n_blocked_status**
+**S2N_BLOCKED_ON_APPLICATION_INPUT**, and will keep giving the same
+error until the **op** is performed on the spawned thread and applied to
+connection. Once the spawned thread finishes performing the **op**,
+the **s2n_async_pkey_op_apply** needs to be called in the main
+event loop thread to apply the result to connection. After that the user
+needs to call **s2n_negotiate** to proceed with handshake.
+
+```c
+typedef int (*s2n_async_pkey_fn)(struct s2n_connection *conn, struct s2n_async_pkey_op *op);
+extern int s2n_config_set_async_pkey_callback(struct s2n_config *config, s2n_async_pkey_fn fn);
+extern int s2n_async_pkey_op_perform(struct s2n_async_pkey_op *op, s2n_cert_private_key *key);
+extern int s2n_async_pkey_op_apply(struct s2n_async_pkey_op *op, struct s2n_connection *conn);
+extern int s2n_async_pkey_op_free(struct s2n_async_pkey_op *op);
+```
+
+- **op** is an opaque object representing private key operation which
+needs to be performed.
+- **key** is a private key used for operation.
+
+**s2n_async_pkey_fn** is invoked every time some action involving
+private key is required. The **conn** provides a pointer to connection
+which triggered this action. The **op** is an operation which needs to be
+performed through **s2n_async_pkey_op_perform** call and applied to
+**conn** through **s2n_async_pkey_op_apply** call after the
+callback was complete. Once performed and applied the **op** should be
+freed by the caller through **s2n_async_pkey_op_free** call.
+
+**s2n_config_set_async_pkey_callback** sets up the callback to invoke
+for asynchronous private key operations and enables asynchronous mode.
+When it is used **s2n_negotiate** will error out with **S2N_ERR_T_BLOCKED**
+error type and **s2n_blocked_status** **S2N_BLOCKED_ON_APPLICATION_INPUT**
+each time the callback was invoked and won't proceed until
+**s2n_async_pkey_op_apply** is called on the connection and
+associated **op** object.
+
+**s2n_async_pkey_op_perform** performs the **op** allowing it to be used
+to resume the handshake through **s2n_async_pkey_op_apply** call. This
+function can be called only once and is safe to call from a different thread.
+To get the private key **key** the caller can use
+**s2n_connection_get_selected_cert** and **s2n_cert_chain_and_key_get_key**
+calls.
+
+**s2n_async_pkey_op_apply** applies the **op** after it was performed
+through **s2n_async_pkey_op_perform** call to connection. The caller will
+need to call **s2n_negotiate** again in order to proceed with TLS handshake.
```

Comment:
Maybe add that this must be performed from same thread that does the main s2n connection interactions? Don't know how to formulate it, since it doesn't really matter from which thread - the main thing you need to guarantee is no other concurrent s2n api calls on the connection.

Reply:
Added note at the beginning that no 2 threads should operate on the same connection at the same time.